File: src/hero_bringup/scripts/lidar_frame_relay.py
#!/usr/bin/env python3
"""Relay Livox point cloud with corrected frame_id for RViz visualization.

The ros_gz_bridge produces a frame_id like 'red_standard_robot1/front_mid360'
(model-scoped), but the TF tree (from robot_state_publisher under namespace
red_standard_robot1) publishes frames as 'front_mid360' (relative).

This relay strips the namespace prefix so RViz can resolve the transform chain:
  chassis -> front_mid360  (from robot_state_publisher URDF)
"""
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy, HistoryPolicy
from sensor_msgs.msg import PointCloud2


class LidarFrameRelay(Node):
    def __init__(self):
        super().__init__('lidar_frame_relay')

        self.declare_parameter('input_topic', '/red_standard_robot1/livox/lidar')
        self.declare_parameter('output_topic', '/red_standard_robot1/livox/lidar_fixed')
        self.declare_parameter('target_frame_id', 'front_mid360')

        input_topic = self.get_parameter('input_topic').value
        output_topic = self.get_parameter('output_topic').value
        self.target_frame_id = self.get_parameter('target_frame_id').value

        # Sub: BEST_EFFORT to match ros_gz_bridge SensorDataQoS
        sub_qos = QoSProfile(
            depth=5,
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
        )
        # Pub: RELIABLE to match RViz PointCloud2 display
        pub_qos = QoSProfile(
            depth=5,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
        )

        self.sub = self.create_subscription(
            PointCloud2, input_topic, self.callback, sub_qos)

        self.pub = self.create_publisher(
            PointCloud2, output_topic, pub_qos)

        self.get_logger().info(
            f'LidarFrameRelay: {input_topic} -> {output_topic} '
            f'(frame_id -> {self.target_frame_id})')

    def callback(self, msg: PointCloud2):
        msg.header.frame_id = self.target_frame_id
        self.pub.publish(msg)


def main():
    rclpy.init()
    node = LidarFrameRelay()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        try:
            rclpy.shutdown()
        except RuntimeError:
            pass  # context already shut down by another node


if __name__ == '__main__':
    main()

File: src/hero_bringup/src/lidar_frame_relay.cpp
/// C++ relay: fix frame_id of Livox pointcloud for RViz TF resolution.
/// Uses BEST_EFFORT throughout — no copies in the critical path.

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<rclcpp::Node>("lidar_frame_relay");

  auto in_topic  = node->declare_parameter("input_topic",  "/red_standard_robot1/livox/lidar");
  auto out_topic = node->declare_parameter("output_topic", "/red_standard_robot1/livox/lidar_fixed");
  auto target    = node->declare_parameter("target_frame_id", "front_mid360");

  rclcpp::QoS qos(rclcpp::KeepLast(10));
  qos.best_effort();
  qos.durability_volatile();

  auto pub = node->create_publisher<sensor_msgs::msg::PointCloud2>(out_topic, qos);

  auto sub = node->create_subscription<sensor_msgs::msg::PointCloud2>(
      in_topic, qos,
      [&pub, &target](sensor_msgs::msg::PointCloud2::ConstSharedPtr msg) {
        auto out = std::make_unique<sensor_msgs::msg::PointCloud2>();
        out->header = msg->header;
        out->header.frame_id = target;
        out->height       = msg->height;
        out->width        = msg->width;
        out->fields       = msg->fields;
        out->is_bigendian = msg->is_bigendian;
        out->point_step   = msg->point_step;
        out->row_step     = msg->row_step;
        out->data         = msg->data;
        out->is_dense     = msg->is_dense;
        pub->publish(std::move(out));
      });

  RCLCPP_INFO(node->get_logger(), "LidarFrameRelay (C++ BEST_EFFORT): %s -> %s (frame: %s)",
              in_topic.c_str(), out_topic.c_str(), target.c_str());
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
